streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+                    builtTopicNodes.get(sinkTopic).addUpStreamTopicsInfo(topicsInfo);
+                } else {
+                    final TopicNode topicNode = new TopicNode(sinkTopic);
+                    topicNode.addUpStreamTopicsInfo(topicsInfo);
+                    builtTopicNodes.put(sinkTopic, topicNode);
                 }
             }
-        } while (numPartitionsNeeded);
+        }
+
+        // 2. Use DFS along with memoization to calc repartition number of all repartitionSourceTopics
+        for (final String topic : allRepartitionSourceTopics) {
+            calcRepartitionNumForTopic(topic, repartitionTopicMetadata, metadata, builtTopicNodes, new HashMap<TopicsInfo, Integer>());
+        }
+    }
+
+    private int calcRepartitionNumForTopic(final String topic,
+                                           final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+                                           final Cluster metadata,
+                                           final Map<String, TopicNode> builtTopicNodes,
+                                           final Map<TopicsInfo, Integer> topicsInfoNumberOfPartitions) {
+
+        if (repartitionTopicMetadata.containsKey(topic)) {
+            final Optional<Integer> maybeNumberPartitions = repartitionTopicMetadata.get(topic).numberOfPartitions();
+            // if numberOfPartitions already calculated, return directly
+            if (maybeNumberPartitions.isPresent()) {
+                return maybeNumberPartitions.get();
+            } else {
+                // else calculate the max numRepartitions of its upStream TopicsInfo and set the repartitionTopicMetadata for memoization before return
+                final TopicNode topicNode = builtTopicNodes.get(topic);
+                Integer maxNumberPartitions = 0;
+                for (final TopicsInfo upstream : topicNode.upStreams) {
+                    final Integer upStreamRepartitionNum = calcRepartitionNumForTopicInfo(upstream, repartitionTopicMetadata,
+                            metadata, builtTopicNodes, topicsInfoNumberOfPartitions);
+                    maxNumberPartitions = upStreamRepartitionNum > maxNumberPartitions ? upStreamRepartitionNum : maxNumberPartitions;
+                }
+                repartitionTopicMetadata.get(topic).setNumberOfPartitions(maxNumberPartitions);
+                return maxNumberPartitions;
+            }
+        } else {
+            final Integer count = metadata.partitionCountForTopic(topic);
+            if (count == null) {
+                throw new IllegalStateException(
+                        "No partition count found for source topic "
+                                + topic
+                                + ", but it should have been."
+                );
+            }
+            return count;
+        }
+    }
+
+    private int calcRepartitionNumForTopicInfo(final TopicsInfo topicsInfo,
```

Comment:
`calcRepartitionNumForTopic` and `calcRepartitionNumForTopicInfo` are pretty hard to differentiate, better to name it more distinct.

Reply:
Yeah, seems to be the case, but in the mutual recursive function, it seems acceptable...